File: src/desktop/src-tauri/src/job.rs
//! Lifecycle of the long model operations: downloading weights, and the
//! FLUX.2-dev conversion.
//!
//! These used to be a `Command` awaited inside a `#[tauri::command]`, with no
//! handle kept anywhere. Three consequences, all observed:
//!
//! * the only guard against starting a second one was a React `disabled`
//!   attribute, and `App` unmounts `Models` on a tab switch — so leaving the tab
//!   and coming back re-armed the button while the first child was still running;
//! * there was no way to stop one, and a FLUX.2-dev conversion runs for hours at
//!   a ~66 GB peak;
//! * `RunEvent::Exit` only reached the *server* child, so quitting the app left
//!   the conversion running as an orphan under launchd, invisible.
//!
//! So the child lives here, in Tauri state, exactly like `Supervisor` holds the
//! server: one slot, signalled through the same process-group ladder. Runtime
//! state only — nothing is persisted, because nothing here survives the process.

use std::sync::Arc;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use serde::Serialize;
use serde_json::Value;
use tokio::process::Child;
use tokio::sync::Mutex;

use crate::supervisor::signal_group;

/// How often the monitor task asks whether the child has exited.
///
/// Reaping under the same lock that cancellation takes is what makes the two
/// race-free, so it has to be a poll rather than a bare `wait()`: `wait()` would
/// own the child and reap it outside the lock, leaving a window where cancel
/// still believes it has a live pid.
const POLL: Duration = Duration::from_millis(250);

/// Grace given to a cancelled child before SIGKILL. Same shape as the server's
/// ladder; shorter, because these children have no in-flight HTTP to drain — the
/// cost of cutting one off is a partial download, which was already the outcome.
const CANCEL_GRACE: Duration = Duration::from_secs(10);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum JobKind {
    Fetch,
    Prequantize,
}

impl JobKind {
    fn label(self) -> &'static str {
        match self {
            JobKind::Fetch => "a model download",
            JobKind::Prequantize => "the FLUX.2-dev conversion",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum JobState {
    #[default]
    Idle,
    Running,
    /// SIGTERM sent, child not yet reaped. Distinct from `Running` so the UI can
    /// stop offering Cancel twice, and from `Cancelled` so it does not claim the
    /// process is gone before it is.
    Cancelling,
    Completed,
    Failed,
    Cancelled,
}

impl JobState {
    fn is_active(self) -> bool {
        matches!(self, JobState::Running | JobState::Cancelling)
    }
}

/// What React needs to reconstruct the operation after `Models` remounts.
#[derive(Debug, Clone, Default, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct JobStatus {
    pub state: JobState,
    pub kind: Option<JobKind>,
    /// Model key for a fetch; the component list for a conversion.
    pub target: Option<String>,
    /// Name of the last structured event on the child's stdout, verbatim.
    pub event: Option<String>,
    /// Its `fields` object, verbatim — we do not reshape the child's schema.
    pub fields: Option<Value>,
    /// Latest human-readable line from the structured stream while running; the
    /// terminal reason once finished.
    pub message: Option<String>,
    pub started_at_ms: Option<u64>,
    pub finished_at_ms: Option<u64>,
}

/// The reason a failed job reports: what the child said on its structured
/// stream, falling back to the exit code only when it said nothing.
///
/// The fallback used to be the whole story — every failure read "the download
/// failed (code Some(1))", whatever had actually gone wrong.
fn failure_message(error_message: Option<&str>, code: Option<i32>) -> String {
    match (error_message, code) {
        (Some(reason), _) => reason.to_owned(),
        (None, Some(code)) => format!("exited with code {code}"),
        (None, None) => "terminated by a signal".to_owned(),
    }
}

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|value| value.as_millis() as u64)
        .unwrap_or(0)
}

#[derive(Default)]
pub struct JobManager {
    /// Held here rather than by the awaiting task on purpose: reading the pid,
    /// signalling it and reaping it then all happen under one lock, so a cancel
    /// can never signal a pid that has already been reaped and reused.
    child: Option<Child>,
    /// Set by `request_cancel`, read when the child is reaped to tell a
    /// cancellation apart from a failure — the exit code cannot, since a
    /// SIGTERM'd Python exits non-zero either way.
    cancelling: bool,
    /// Last ERROR-level message from the structured stream, kept as the terminal
    /// reason candidate. Exit codes alone said "code Some(1)" and nothing else.
    error_message: Option<String>,
    status: JobStatus,
}

pub type SharedJobs = Arc<Mutex<JobManager>>;

impl JobManager {
    /// Adopt a freshly spawned child. The caller has already checked that no
    /// other job is active, under the same lock.
    pub fn begin(&mut self, kind: JobKind, target: String, child: Child) {
        self.child = Some(child);
        self.cancelling = false;
        self.error_message = None;
        self.status = JobStatus {
            state: JobState::Running,
            kind: Some(kind),
            target: Some(target),
            started_at_ms: Some(now_ms()),
            ..Default::default()
        };
    }

    /// `Err` when a job is already active, naming it so the message is actionable.
    pub fn ensure_free(&self) -> Result<(), String> {
        if self.status.state.is_active() {
            let what = self.status.kind.map(JobKind::label).unwrap_or("an operation");
            let target = self.status.target.clone().unwrap_or_default();
            return Err(format!(
                "{what}{} is already running. Only one heavy model operation runs at a time — \
                 they compete for the same unified memory and the same HuggingFace cache. \
                 Wait for it, or cancel it first.",
                if target.is_empty() {
                    String::new()
                } else {
                    format!(" ({target})")
                }
            ));
        }
        Ok(())
    }

    pub fn status(&mut self) -> JobStatus {
        self.reap();
        self.status.clone()
    }

    /// Notice a child that has exited and settle the terminal state.
    fn reap(&mut self) {
        let Some(child) = self.child.as_mut() else {
            return;
        };
        // `None` = still running, `Err` = status unreadable; both leave it alone.
        let Ok(Some(exit)) = child.try_wait() else {
            return;
        };
        self.child = None;
        self.status.finished_at_ms = Some(now_ms());
        if self.cancelling {
            self.status.state = JobState::Cancelled;
            self.status.message = Some("Cancelled.".to_owned());
        } else if exit.success() {
            self.status.state = JobState::Completed;
            self.status.message = None;
        } else {
            self.status.state = JobState::Failed;
            self.status.message = Some(failure_message(self.error_message.as_deref(), exit.code()));
        }
        self.cancelling = false;
    }

    /// SIGTERM the child's process group. `Err` when nothing is running.
    ///
    /// Returns the pid the SIGKILL escalation must target, read in this same
    /// critical section: fetching it in a second lock would let a job that
    /// finished meanwhile hand the escalation to whatever started next.
    pub fn request_cancel(&mut self) -> Result<(JobStatus, u32), String> {
        self.reap();
        if !self.status.state.is_active() {
            return Err("No model operation is running.".to_owned());
        }
        // Reading the pid under the same lock as `reap` is what keeps this from
        // ever signalling a reaped, recycled pid.
        let Some(pid) = self.child.as_ref().and_then(Child::id) else {
            return Err("The operation has already finished.".to_owned());
        };
        self.cancelling = true;
        self.status.state = JobState::Cancelling;
        self.status.message = Some("Stopping…".to_owned());
        signal_group(pid, libc::SIGTERM);
        Ok((self.status.clone(), pid))
    }

    /// Escalate a cancellation that SIGTERM did not finish. No-op if the child
    /// already exited or a different job has since started.
    fn escalate(&mut self, pid: u32) {
        self.reap();
        if !self.status.state.is_active() {
            return;
        }
        if self.child.as_ref().and_then(Child::id) == Some(pid) {
            signal_group(pid, libc::SIGKILL);
        }
    }

    /// Terminate without waiting, for application exit.
    pub fn kill_now(&mut self) {
        if let Some(pid) = self.child.as_ref().and_then(Child::id) {
            signal_group(pid, libc::SIGKILL);
        }
        self.child = None;
    }

    /// Fold one structured stdout line into the status.
    ///
    /// Only the fields the UI actually needs: the event name and its `fields` for
    /// progress, `message` for the current step, and an ERROR-level `message` as
    /// the terminal reason. Everything else on the line is ignored, and the raw
    /// line still goes to the Logs tab untouched — this is not a second protocol.
    fn note_line(&mut self, line: &str) {
        if !self.status.state.is_active() {
            return;
        }
        let Ok(Value::Object(record)) = serde_json::from_str::<Value>(line) else {
            return;
        };
        if let Some(event) = record.get("event").and_then(Value::as_str) {
            self.status.event = Some(event.to_owned());
            self.status.fields = record.get("fields").cloned();
        }
        if let Some(message) = record.get("message").and_then(Value::as_str) {
            self.status.message = Some(message.to_owned());
            if matches!(
                record.get("level").and_then(Value::as_str),
                Some("ERROR") | Some("CRITICAL")
            ) {
                self.error_message = Some(message.to_owned());
            }
        }
    }
}

/// Record a structured line against the active job, if any.
pub async fn note_line(jobs: &SharedJobs, line: &str) {
    jobs.lock().await.note_line(line);
}

/// Drive the job to a terminal state without waiting on React to poll.
///
/// A plain `child.wait()` would own the child outside the lock; polling
/// `try_wait` under it keeps pid handling in one place — see `POLL`.
pub fn monitor(jobs: SharedJobs) {
    tauri::async_runtime::spawn(async move {
        loop {
            tokio::time::sleep(POLL).await;
            let mut guard = jobs.lock().await;
            guard.reap();
            if !guard.status.state.is_active() {
                return;
            }
        }
    });
}

/// SIGTERM → bounded wait → SIGKILL, the same ladder the server shutdown uses.
pub fn arm_kill_after_grace(jobs: SharedJobs, pid: u32) {
    tauri::async_runtime::spawn(async move {
        tokio::time::sleep(CANCEL_GRACE).await;
        jobs.lock().await.escalate(pid);
    });
}

#[cfg(test)]
mod tests {
    use super::*;

    /// A manager that believes a job is running, without a real child. Enough to
    /// exercise the line folding and the single-flight check, which are the parts
    /// with logic in them.
    fn running() -> JobManager {
        JobManager {
            status: JobStatus {
                state: JobState::Running,
                kind: Some(JobKind::Fetch),
                target: Some("z-image-turbo".to_owned()),
                ..Default::default()
            },
            ..Default::default()
        }
    }

    #[test]
    fn structured_error_becomes_the_terminal_reason() {
        // The point of parsing the stream at all: "code Some(1)" told the user
        // nothing about *why*.
        let mut jobs = running();
        jobs.note_line(r#"{"ts":"t","level":"INFO","logger":"l","message":"Fetching z-image-turbo"}"#);
        jobs.note_line(r#"{"ts":"t","level":"ERROR","logger":"l","message":"Unknown model 'nope'"}"#);
        assert_eq!(jobs.error_message.as_deref(), Some("Unknown model 'nope'"));
        assert_eq!(failure_message(jobs.error_message.as_deref(), Some(1)), "Unknown model 'nope'");
    }

    #[test]
    fn without_a_structured_reason_the_exit_code_is_the_fallback() {
        assert_eq!(failure_message(None, Some(1)), "exited with code 1");
        assert_eq!(failure_message(None, None), "terminated by a signal");
    }

    #[test]
    fn an_info_line_updates_the_message_but_is_not_a_failure_reason() {
        let mut jobs = running();
        jobs.note_line(r#"{"level":"INFO","message":"Downloading / reading repo/x"}"#);
        assert_eq!(jobs.status.message.as_deref(), Some("Downloading / reading repo/x"));
        assert_eq!(jobs.error_message, None);
    }

    #[test]
    fn progress_events_are_kept_verbatim() {
        let mut jobs = running();
        jobs.note_line(
            r#"{"level":"INFO","message":"  quantized 8/56 blocks","event":"prequantize_progress","fields":{"block":8,"blocks":56}}"#,
        );
        assert_eq!(jobs.status.event.as_deref(), Some("prequantize_progress"));
        let fields = jobs.status.fields.as_ref().expect("fields kept");
        assert_eq!(fields.get("block").and_then(Value::as_u64), Some(8));
        assert_eq!(fields.get("blocks").and_then(Value::as_u64), Some(56));
    }

    #[test]
    fn non_json_output_is_ignored_rather_than_guessed_at() {
        // tqdm fragments and tracebacks share this stream's sibling; nothing here
        // may try to read meaning out of human text.
        let mut jobs = running();
        jobs.note_line("Traceback (most recent call last):");
        jobs.note_line("  4%|##        | 1.2G/28G [00:31<11:20, 39.4MB/s]");
        assert_eq!(jobs.status.message, None);
        assert_eq!(jobs.error_message, None);
    }

    #[test]
    fn a_finished_job_stops_absorbing_late_lines() {
        // Pumps outlive the child briefly; a straggler must not overwrite the
        // terminal reason that has already been settled.
        let mut jobs = running();
        jobs.status.state = JobState::Failed;
        jobs.status.message = Some("Unknown model 'nope'".to_owned());
        jobs.note_line(r#"{"level":"INFO","message":"late line"}"#);
        assert_eq!(jobs.status.message.as_deref(), Some("Unknown model 'nope'"));
    }

    #[test]
    fn single_flight_is_enforced_and_names_what_is_running() {
        let jobs = running();
        let refused = jobs.ensure_free().expect_err("a second job must be refused");
        assert!(refused.contains("a model download"), "names the kind: {refused}");
        assert!(refused.contains("z-image-turbo"), "names the target: {refused}");

        assert!(JobManager::default().ensure_free().is_ok());
    }

    #[test]
    fn a_terminal_job_no_longer_blocks_the_next_one() {
        let mut jobs = running();
        jobs.status.state = JobState::Cancelled;
        assert!(jobs.ensure_free().is_ok());
    }
}

File: src/dashboard/src/job.ts
/**
 * The server-owned long operation, polled once for the whole page.
 *
 * The server is the authority: a download or a conversion is a child of the
 * server process, not of a React component, and survives the panel that started
 * it — and now the browser tab too. The poll lives here, above the views, for
 * one reason the panels cannot provide themselves: switching to Logs and back
 * used to unmount `Models`, stop its poll, and re-enter with no operation on
 * screen until the next tick. The state a user walks away from has to be the
 * state they walk back to.
 *
 * Nothing here decides anything. It reads a status and hands it on.
 */
import { useCallback, useEffect, useRef, useState } from "react";

import * as api from "./api";
import { messageOf } from "./api";
import type { JobStatus } from "./types";

/** How often we ask the server what the long operation is doing. */
const POLL_MS = 1000;

export function isActive(job: JobStatus | null): boolean {
  return job?.state === "running" || job?.state === "cancelling";
}

/**
 * One line describing what the operation is doing, from its structured stream.
 *
 * The child's own event names and fields, verbatim — this maps them to English
 * and invents nothing. A percentage is deliberately absent: `mflux_save` writes
 * a model in one call and reports phases, so a fraction here would be a number
 * we made up.
 */
export function describeJob(job: JobStatus): string {
  const fields = job.fields ?? {};
  if (job.event === "prequantize_progress") {
    const { block, blocks } = fields as { block?: number; blocks?: number };
    if (block && blocks) return `quantizing block ${block} of ${blocks}`;
  }
  if (job.event === "prequantize_component_start") {
    const { component } = fields as { component?: string };
    if (component) return `reading ${component}`;
  }
  if (job.event === "prequantize_component_done") {
    const { component, written_gb } = fields as { component?: string; written_gb?: number };
    if (component) return `wrote ${component}${written_gb ? ` (${written_gb} GB)` : ""}`;
  }
  return job.message ?? "starting…";
}

/**
 * What a finished operation actually achieved, from the child's own last word.
 *
 * The distinction this exists to preserve: a conversion run can succeed without
 * producing anything usable. `prequantize_done` is emitted only after Python has
 * validated every required component and written the completion marker;
 * `prequantize_partial` is emitted when the run did what was asked and the
 * artifact is still incomplete. Both are successful exits, and telling them
 * apart from an exit code is impossible — which is why nothing here looks at
 * one.
 *
 * `labelFor` translates a component key into the name the backend publishes for
 * it, so this reports "Text encoder" rather than `text_encoder` without keeping
 * a list of its own.
 */
export function describeOutcome(
  job: JobStatus,
  labelFor: (key: string) => string = (key) => key,
): string | null {
  const fields = job.fields ?? {};
  if (job.event === "prequantize_done") {
    const { bits } = fields as { bits?: number };
    return bits ? `${bits}-bit variant ready and selected.` : "Variant ready and selected.";
  }
  if (job.event === "prequantize_partial") {
    const { completed, missing } = fields as { completed?: string[]; missing?: string[] };
    const done = (completed ?? []).map(labelFor);
    const left = (missing ?? []).length;
    if (!done.length) return null;
    return (
      `${done.join(", ")} converted - ` +
      `${left} component${left === 1 ? "" : "s"} remaining before this variant can be used.`
    );
  }
  return null;
}

export type JobView = {
  job: JobStatus | null;
  /** The server could not be asked. Not the same as "no job is running". */
  error: string | null;
  active: boolean;
  /** Read it again now, rather than waiting for the next tick. */
  refresh: () => Promise<void>;
  /** Fires when a job reaches a terminal state, so views can reload what it changed. */
  onSettled: (listener: (job: JobStatus) => void) => () => void;
};

export function useJob(): JobView {
  const [job, setJob] = useState<JobStatus | null>(null);
  const [error, setError] = useState<string | null>(null);
  const previous = useRef<JobStatus | null>(null);
  const listeners = useRef(new Set<(job: JobStatus) => void>());

  const refresh = useCallback(async () => {
    try {
      const status = await api.jobStatus();
      setJob(status);
      setError(null);
      const before = previous.current;
      // The edge, not the level: a completed job stays completed until the next
      // one starts, and firing on every tick would reload the catalogue forever.
      if (before && isActive(before) && !isActive(status)) {
        for (const listener of listeners.current) listener(status);
      }
      previous.current = status;
    } catch (cause) {
      setError(messageOf(cause));
    }
  }, []);

  useEffect(() => {
    void refresh();
    const timer = setInterval(() => void refresh(), POLL_MS);
    return () => clearInterval(timer);
  }, [refresh]);

  const onSettled = useCallback((listener: (job: JobStatus) => void) => {
    listeners.current.add(listener);
    return () => {
      listeners.current.delete(listener);
    };
  }, []);

  return { job, error, active: isActive(job), refresh, onSettled };
}
